docs/design/diagnostics.md

```diff
@@ -0,0 +1,293 @@
+
+type Diagnostics []Diagnostic
+
+func (diags Diagnostics) HasError() bool
+
+func Errorf(format string, a ...interface{}) Diagnostics
+
+func FromErr(err error) Diagnostics
+```
+
+## Goals
+
+Diagnostics in this framework should be:
+
+- Able to natively support warning and error diagnostics.
+- Implementable by provider developers.
+- Abstracted from terraform-plugin-go and convertable into those types to separate implementation concerns.
+- Ergonomic to create, read, and delete in Go code (e.g. have helper methods for common use cases).
+
+Additional consideration should be given to:
+
+- Extensibility and additional data storage (e.g. distinguishing between types of diagnostics or what generated an error).
+- The ability to deduplicate or manipulate diagnostics.
+
+## Proposals
+
+### Thin Abstraction
+
+The most trivial implementation would be creating new Go types in this framework that provide a thin abstraction layer, e.g.
+
+```go
+package diag
+
+type Diagnostic struct {
+    Severity  DiagnosticSeverity
+    Summary   string
+    Detail    string
+    Attribute *tftypes.AttributePath // See also: https://github.com/hashicorp/terraform-plugin-framework/issues/81
+}
+
+type Diagnostics []Diagnostic
+
+func (d Diagnostics) Append(in ...Diagnostic)
+func (d Diagnostics) HasError() bool
+func (d Diagnostics) toTfprotov6Diagnostics() []*tfprotov6.Diagnostic
+```
+
+This solves many requirements, allowing the framework to own the abstraction and provide ergonomic methods. However, this implementation is tightly coupled to the underlying implementation and does not introduce protections against backwards compatibility issues. Since the data fields are strongly associated with the type, implementors inside and outside the framework are bound to the initial implementation details. If those details are not exported and instead hidden behind a constructor function, e.g.
+
+```go
+type Diagnostic struct {
+    severity  DiagnosticSeverity
+    summary   string
+    detail    string
+    attribute *tftypes.AttributePath // See also: https://github.com/hashicorp/terraform-plugin-framework/issues/81
+}
+
+func NewDiagnostic(severity DiagnosticSeverity, summary string, detail string, attribute *tftypes.AttributePath) Diagnostic
+func NewErrorDiagnostic(summary string, detail string, attribute *tftypes.AttributePath) Diagnostic
+func NewWarningDiagnostic(summary string, detail string, attribute *tftypes.AttributePath) Diagnostic
+```
+
+This type of implementation will still be difficult to extend or change.
+
+As another extensibility example, as `Diagnostics` pass through various functions across subsystems, determining the "source" or "cause" of a diagnostic must be done by `Summary` and/or `Detail` value similar to checking Go `error` types by value using `(error).Is()`. Extending the `Diagnostic` type is possible to add other identifying fields or creating new types, e.g.
+
+```go
+type ValidationDiagnostic struct {
+  Diagnostic
+}
+```
```

Comment:
I don't know that this would work. I think we'd get compilation errors on this:

```go
func doTheThing() Diagnostic {
  return ValidationDiagnostic{}
}
```

I don't think that would work 🤔 I think that kind of embedding only works for interfaces; the struct embedding I think just inherits properties and methods, but the two types can't be used interchangeably. (I'd love to be corrected if I'm wrong, or if I'm misunderstanding what you're suggesting here.)

Reply:
Fixed, thanks for calling that out.